k4FWCore/include/k4FWCore/DataHandle.h

```diff
@@ -71,10 +69,18 @@ DataHandle<T>::DataHandle(const std::string& descriptor, Gaudi::DataHandle::Mode
     : DataObjectHandle<DataWrapper<T>>(descriptor, a, fatherAlg), m_eds("EventDataSvc", "DataHandle") {
   if (a == Gaudi::DataHandle::Writer) {
     StatusCode    sc = m_eds.retrieve();
-    PodioDataSvc* pds;
-    pds       = dynamic_cast<PodioDataSvc*>(m_eds.get());
     m_dataPtr = nullptr;
-    if (nullptr != pds) {
+    PodioDataSvc* podio_data_service;
+
+    podio_data_service = dynamic_cast<PodioDataSvc*>(m_eds.get());
+    if (nullptr != podio_data_service) {
+      m_dataPtr = new T();
+    } else {
+
+    // This is the legacy implementation kept for a transition period
+    PodioLegacyDataSvc* plds;
+    plds       = dynamic_cast<PodioLegacyDataSvc*>(m_eds.get());
+    if (nullptr != plds) {
       if (std::is_convertible<T*, podio::CollectionBase*>::value) {
```

Comment:
```suggestion
      if constexpr (std::is_convertible<T*, podio::CollectionBase*>::value) {
```
This should fix the `k4Gen` issue